ldbreclaim/ldbreclaim.go

```diff
@@ -0,0 +1,287 @@
+}
+
+// Len is part of sort.Interface.
+func (s *ldbSorter) Len() int {
+	return len(s.ldbs)
+}
+
+// Swap is part of sort.Interface.
+func (s *ldbSorter) Swap(i, j int) {
+	s.ldbs[i], s.ldbs[j] = s.ldbs[j], s.ldbs[i]
+}
+
+// Less is part of sort.Interface. It is implemented by calling the "by" closure in the sorter.
+func (s *ldbSorter) Less(i, j int) bool {
+	return s.by(&s.ldbs[i], &s.ldbs[j])
+}
+
+var _ldbs []Ldb
+var totalLdbs int64 = 0
+var totalSize int64 = 0
+var totalKeys int64 = 0
+var totalDeletedKeys int64 = 0
+var averageKeySize int64 = 0
+
+func DumpStats() {
+	fmt.Printf("totalLdbs=%v\n", totalLdbs)
+	fmt.Printf("totalSize=%v bytes\n", totalSize)
+	fmt.Printf("totalKeys=%v bytes\n", totalKeys)
+	fmt.Printf("totalDeletedKeys=%v\n", totalDeletedKeys)
+	fmt.Printf("averageKeySize=%v\n", averageKeySize)
+}
+
+func LdbAdd(ldb Ldb) {
+	_ldbs = append(_ldbs, ldb)
+	totalLdbs++
+	totalSize += ldb.size
+	totalKeys += ldb.numKeys
+	totalDeletedKeys += ldb.numDeletedKeys
+}
+
+func ComputeAverageKeySize() {
+	averageKeySize = totalSize / totalKeys
+}
+
+func LdbSortByStartKey() {
+	// sort keys
+	startKey := func(p1, p2 *Ldb) bool {
+		return p1.startKey < p2.startKey
+	}
+	By(startKey).Sort(_ldbs)
+}
+
+func LdbSortByNumKeys() {
+	// sort keys
+	numKeys := func(p1, p2 *Ldb) bool {
+		return p1.numKeys > p2.numKeys
+	}
+	By(numKeys).Sort(_ldbs)
+}
+
+// find the first ldb that starts after or at prefix
+func LdbFindStartKey(prefix string) int {
+	startKey := func(i int) bool {
+		return _ldbs[i].startKey >= prefix
+	}
+	return sort.Search(len(_ldbs), startKey)
+}
+
+/*
+func LdbDump(s []Ldb) {
+	for i, ldb := range s {
+		fmt.Printf("%v %v\n", i, ldb)
+	}
+}
+*/
+
+// remove ranges when endKey is < key
+func RemoveRangesBefore(s []Ldb, key string) []Ldb {
+	tmp := s[:0]
+	for _, ldb := range s {
+		if ldb.endKey >= key {
+			tmp = append(tmp, ldb)
+		}
+	}
+	return tmp
+}
+
+func AddToRanges(s []Ldb, numKeys int64, numDeletedKeys int64) {
+	for _, ldb := range s {
+		_ldbs[ldb.index].numKeys += numKeys
+		_ldbs[ldb.index].numDeletedKeys += numDeletedKeys
+	}
+}
+
+func FindRanges() {
+	var currentRanges []Ldb
+
+	for i, ldb := range _ldbs {
+		currentRanges = RemoveRangesBefore(currentRanges, ldb.startKey)
+		AddToRanges(currentRanges, ldb.numKeys, ldb.numDeletedKeys)
+		newLdb := ldb
+		newLdb.index = i
+		currentRanges = append(currentRanges, newLdb)
+	}
+}
+
+// call after being sorted by numKeys
+func DumpRanges() {
+	var resultsFound int = 0
+	for _, ldb := range _ldbs {
+		if *verbose {
+			fmt.Printf("%v\n", ldb)
+		}
+		worstCaseSpaceNeededMb := (ldb.numKeys * averageKeySize) / (1024 * 1024)
+		spaceReclaimedMb := (ldb.numDeletedKeys * averageKeySize) / (1024 * 1024)
+		if spaceReclaimedMb > *minReclaimMb &&
+			worstCaseSpaceNeededMb < *tempSpaceMb {
+			fmt.Printf("%v\n", ldb)
+			resultsFound++
+		}
+		if resultsFound >= *numResults {
+			break
+		}
+	}
+}
+
+func Kind(k []byte) uint8 {
+        return k[len(k)-8]
```

Comment:
Could be interesting to have hints on the logic at play here, like a docstring for the `Kind` function and what is placed at `len(k)-8` or the general format of a key